File: 다트 게임.py
def pow(n, p):
    if p == 'S':
        return n
    elif p == 'D':
        return n * n
    elif p == 'T':
        return n * n * n

def solution(dartResult):
    answer = 0
    idx = -1
    a = [0 for _ in  range(3)]
    size = len(dartResult)

    i = 0
    while i < size:
        cur = dartResult[i]

        if cur == 'S' or cur == 'D' or cur == 'T':
            a[idx] = pow(a[idx], cur)
        elif cur == '*':
            if idx != 0:
                a[idx - 1] *= 2
            a[idx] *= 2
        elif cur == '#':
            a[idx] *= -1
        else:
            n = ord(cur) - ord('0')
            if cur == '1' and dartResult[i + 1] == '0':
                n = 10
                i += 1
            idx += 1
            a[idx] = n
        i += 1
    for i in range(3):
        print(a[i])
        answer += a[i]
    return answer

print(solution("1S2D*3T"))

File: 다트 게임.cpp
#include <string>
#include <vector>
#include <cmath>
#include <iostream>
using namespace std;

int check_opt(string dartResult, int i)
{
	if (i == dartResult.length())
		return 1;
	else if (dartResult[i + 1] == '*' || dartResult[i + 1] == '#')
		return (dartResult[i + 1] == '*') ? 2 : -1;
	else
		return 1;
}
int solution(string dartResult) {
	int answer = 0;
	int current = 0;
	int p = 0;
	int score[3]{ 0, 0, 0 };
	char opt;
	for (int i = 0; i < dartResult.length(); i++)
	{
		if (dartResult[i] == 'S') p = 1;
		else if (dartResult[i] == 'D') p = 2;
		else if (dartResult[i] == 'T') p = 3;
		if (p)
		{
			int tmp = stoi(dartResult.substr(0, i));
			opt = check_opt(dartResult, i);
			if (opt == 2 && current)
				score[current - 1] *= 2;
			score[current] = pow(tmp, p) * opt;
			if (opt != 1) dartResult = dartResult.substr(i + 2);
			else dartResult = dartResult.substr(i + 1);
			current++;
			i = 0;
			p = 0;
		}
	}
	answer = score[0] + score[1] + score[2];
	return answer;
}
